### PR TITLE
Enhance utility methods and get rid of normalization processor in beats.

### PR DESCRIPTION
**Pull request by simitt**
This requires following change in beats: https://github.com/elastic/beats/pull/6419

With this PR we get rid of using the beats normalizer, but do all the necessary normalization steps already in the `utility.MapStrEnhancer`. 

implements https://github.com/elastic/apm-server/issues/711

### CODE REVIEW

File: model/system.go [hunk 1]
```diff
@@ -22,5 +22,8 @@ func (s *System) Transform() common.MapStr {
 	utility.Add(system, "platform", s.Platform)
 	utility.Add(system, "ip", s.IP)
 
+	if len(system) == 0 {
```
Comment (jalvz) [inline model/system.go]: i don't think you need this, `utility.Add` is already handling empty `MapStr`s

File: processor/transaction/span_context.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+func (c *SpanContext) Transform(m common.MapStr) common.MapStr {
+	if m == nil {
+		m = common.MapStr{}
+	} else {
+		for k, v := range m {
+			utility.Add(m, k, v)
+		}
+	}
+	utility.Add(m, "service", c.service)
+	if len(m) == 0 {
```
Comment (jalvz) [inline processor/transaction/span_context.go]: same as before, this should be needed

File: model/context.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	} else {
+		for k, v := range m {
+			utility.Add(m, k, v)
+		}
+	}
+	utility.Add(m, "service", c.service)
+	utility.Add(m, "process", c.process)
+	utility.Add(m, "system", c.system)
+	utility.MergeAdd(m, "user", c.user)
+	if len(m) == 0 {
```
Comment (jalvz) [inline model/context.go]: this shouldnt be needed

File: model/context.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		system:  system.Transform(),
+		user:    user,
+	}
+}
+
+func (c *Context) Transform(m common.MapStr) common.MapStr {
+	if m == nil {
+		m = common.MapStr{}
+	} else {
+		for k, v := range m {
```
Comment (jalvz) [inline model/context.go]: this reads weird at first sight, can you add a short comment saying that you do this to remove nulls? same for spanContext

File: utility/map_str_enhancer.go [hunk 5]
```diff
@@ -4,44 +4,99 @@ import (
...
 		}
 	case []common.MapStr:
 		if valArr := val.([]common.MapStr); len(valArr) > 0 {
 			m[key] = valArr
+		} else {
+			delete(m, key)
 		}
+	case float64:
+		m[key] = common.Float(val.(float64))
 	default:
```
Comment (jalvz) [inline utility/map_str_enhancer.go]: beats normalizer also calls `UTC()` on time objects

File: processor/transaction/span_context.go (hunk 2 above)
Comment (simitt) [reply processor/transaction/span_context.go]: It is necessary as it checks whether ot not `m` has elements, if this check is removed an empty hash would be returned instead of `nil` and stored in ES.

File: utility/map_str_enhancer.go (hunk 5 above)
Comment (simitt) [reply utility/map_str_enhancer.go]: This is done when decoding time objects, https://github.com/simitt/apm-server/blob/ad8d5d26d243e955cfcb252852a864ee7dac95cb/utility/rfc3399decoder.go#L15. I think this should be enough.

File: processor/transaction/span_context.go (hunk 2 above)
Comment (jalvz) [reply processor/transaction/span_context.go]: why? aren't you already doing `if len(valMap) > 0 `  in `utility.Add`?

File: utility/map_str_enhancer.go [hunk 6]
```diff
@@ -4,44 +4,99 @@ import (
...
+			if len(valMap) > 0 {
+				m[key] = valMap
+			} else {
+				delete(m, key)
+			}
+		} else {
+			delete(m, key)
+		}
+	case map[string]interface{}:
+		valMap := val.(map[string]interface{})
```
Comment (jalvz) [inline utility/map_str_enhancer.go]: this wouldn't handle something like `{"a: {"b": null}}`, right?

File: utility/map_str_enhancer.go [hunk 7]
```diff
@@ -4,44 +4,99 @@ import (
...
 	default:
 		if val != nil {
 			m[key] = val
+		} else {
+			delete(m, key)
+		}
+	}
+}
+
+// MergeAdd modifies `data` *in place*, inserting `values` at the given `key`.
```
Comment (jalvz) [inline utility/map_str_enhancer.go]: wrong arg names in comment

File: utility/map_str_enhancer.go [hunk 8]
```diff
@@ -51,3 +106,14 @@ func MillisAsMicros(ms float64) common.MapStr {
 	m["us"] = int(ms * 1000)
 	return m
 }
+
+func AddStrWithDefault(m common.MapStr, key string, val *string, defaultVal string) {
```
Comment (jalvz) [inline utility/map_str_enhancer.go]: this is not used anywhere?

File: utility/rfc3399decoder.go [hunk 9]
```diff
@@ -8,7 +8,11 @@ import (
 func RFC3339DecoderHook(f reflect.Type, t reflect.Type, data interface{}) (interface{}, error) {
 	// stolen from https://github.com/mitchellh/mapstructure/issues/41
 	if t == reflect.TypeOf(time.Time{}) && f == reflect.TypeOf("") {
-		return time.Parse(time.RFC3339, data.(string))
+		parsed, err := time.Parse(time.RFC3339, data.(string))
+		if err != nil {
+			return nil, err
+		}
+		return time.Time(parsed.UTC()), nil
```
Comment (jalvz) [inline utility/rfc3399decoder.go]: ah, i see now you added the UTC call here. forget the previous comment then

File: utility/map_str_enhancer.go (hunk 8 above)
Comment (simitt) [reply utility/map_str_enhancer.go]: This is not a new function, I only moved it, but you're right, so I'll remove it.

**Comment by jalvz** [comment]
I think would be good to copy over the changes of beats #6419 here so this compiles and pass the tests - they will be wiped after rebasing anyways.
That way we could also do some performance tests on this as is, I think is going to be a very noticeable jump.

File: utility/map_str_enhancer.go (hunk 7 above)
Comment (jalvz) [reply utility/map_str_enhancer.go]: as a reminder, if you don't replace the maps with `MapStr` in the decoding in this PR (which is ok, this already big), then I think we have to do it as soon as this is merged - is not great to have duplicated functions like this.

File: processor/transaction/span_context.go (hunk 2 above)
Comment (simitt) [reply processor/transaction/span_context.go]: `m` is the base map to which everything is added. If nothing is added `nil` should be returned.

File: utility/map_str_enhancer.go (hunk 6 above)
Comment (simitt) [reply utility/map_str_enhancer.go]: It should, as I iterate through all the elements (see below). 
I added a spec to show this case though.

File: processor/transaction/span_context.go (hunk 2 above)
Comment (jalvz) [reply processor/transaction/span_context.go]: ok, i don't get it. i commented out all the `if len(m) == 0` in this PR, and all the approval tests pass. Can you add an approval test that exercise this code path? I don't understand how affects the final result.

**Comment by simitt** [comment]
I did have it copied locally and did some performance improvements - the beats PR is approved, so we should have it in soon.

File: CHANGELOG.asciidoc [hunk 10]
```diff
@@ -17,6 +17,7 @@ https://github.com/elastic/apm-server/compare/4daa36bd5c144cf9182afc62dc8042af66
 ==== Added
 
 - Enriched data with IP and UserAgent {pull}393[393].
+- Remove usage of beats normalizer for beat.Events {pull}687[687]
```
Comment (jalvz) [inline CHANGELOG.asciidoc]: this is an implementation detail, i don't think is something users need to be aware of.

File: utility/map_str_enhancer.go (hunk 6 above)
Comment (jalvz) [reply utility/map_str_enhancer.go]: ah ofc 🤦‍♂️ sorry

File: processor/transaction/span_context.go (hunk 2 above)
Comment (simitt) [reply processor/transaction/span_context.go]: you're right that this is checked when the `context` is added to the events, removed it.

**Comment by jalvz** [comment]
thanks for https://github.com/elastic/apm-server/pull/687/commits/bf20be3c42054ee656ba6d7dc26466ad7738f09b
i could reproduce ~15% improvement, which was expected.

**Comment by simitt** [comment]
I'd like to dig a little bit more into how the data parts that do not go through a Go struct are stored in ES before merging this.

**Comment by simitt** [comment]
@elastic/apm-server this is ready for final review.

**Comment by roncohen** [comment]
@simitt I'm investigating something at the moment, please hold on merging

**Comment by roncohen** [comment]
with #709 in I'm OK to merge this

**Comment by simitt** [comment]
@roncohen I've rebased to also include the ES integration test now, as you can see one empty hash is removed, otherwise it stays the same. (see https://github.com/elastic/apm-server/pull/687/files#diff-f63f8f194850ee66420e85c23183a538R122)

**Comment by simitt** [comment]
jenkins, retest this please
